File: 3. Behavioral Design Patterns/Memento/Typescript/MementoTemplate.ts
// Memento Pattern – TypeScript Template

class Memento {
  constructor(private state: string) {}
  getState(): string { return this.state }
}

class Originator {
  private state = ''
  setState(state: string): void { this.state = state }
  getState(): string { return this.state }
  createMemento(): Memento { return new Memento(this.state) }
  restoreMemento(memento: Memento): void { this.state = memento.getState() }
}

class Caretaker {
  private mementos: Memento[] = []
  saveState(originator: Originator): void { this.mementos.push(originator.createMemento()) }
  restoreState(originator: Originator): void { if (this.mementos.length > 0) originator.restoreMemento(this.mementos.pop()!) }
}

;(function main() {
  const originator = new Originator()
  const caretaker = new Caretaker()
  originator.setState('State1')
  caretaker.saveState(originator)
  originator.setState('State2')
  caretaker.restoreState(originator)
})()

export {}

File: 3. Behavioral Design Patterns/Memento/Python/MementoTemplate.py
# Memento Pattern – Python Template

class Memento:
    def __init__(self, state):
        self._state = state
    
    def get_state(self):
        return self._state

class Originator:
    def __init__(self):
        self._state = ""
    
    def set_state(self, state):
        self._state = state
    
    def get_state(self):
        return self._state
    
    def create_memento(self):
        return Memento(self._state)
    
    def restore_memento(self, memento):
        self._state = memento.get_state()

class Caretaker:
    def __init__(self):
        self._mementos = []
    
    def save_state(self, originator):
        self._mementos.append(originator.create_memento())
    
    def restore_state(self, originator):
        if self._mementos:
            memento = self._mementos.pop()
            originator.restore_memento(memento)

def main():
    originator = Originator()
    caretaker = Caretaker()
    
    originator.set_state("State1")
    caretaker.save_state(originator)
    
    originator.set_state("State2")
    caretaker.restore_state(originator)
    print(f"Restored state: {originator.get_state()}")

if __name__ == "__main__":
    main()
